### PR TITLE
Speed up command line help

### PR DESCRIPTION
**Pull request by TomekTrzeciak**
Speed up command line help by avoiding importing heavy modules and reinstate Google docstring processing after #1101 broke it.

| command | master[*] | this branch |
|-|:-:|:-:|
| `improver help` | 4.438s | 2.977s |
| `improver help wxcode` | 2.733s | 1.257s |

[*] after applying fix to `clizefy` from this PR.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for fixed bug

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity decreasing per file
==============================
+ improver/wxcode/utilities.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025&bid=16191239)

File: improver/utilities/compare.py [hunk 1]
```diff
@@ -47,9 +47,11 @@
 import netCDF4
 import numpy as np
 
-TIGHT_TOLERANCE = 1e-5
-DEFAULT_TOLERANCE = 1e-4
-LOOSE_TOLERANCE = 1e-3
+from improver.constants import (
```
Comment (benfitzpatrick) [inline improver/utilities/compare.py]: Codacy Issue found: [Unused LOOSE_TOLERANCE imported from improver.constants](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

File: improver/wxcode/utilities.py [hunk 2]
```diff
@@ -200,35 +198,22 @@ def interrogate_decision_tree(wxtree):
...
-                                  return_index=True)
-        output.append('{}; thresholds: {}'.format(
-            requirement, ', '.join([
-                '{} ({})'.format(str(threshold.points.item()),
-                                 str(threshold.units))
-                for threshold, relation in
-                zip(entries[thresholds], relations[thresholds])])))
+    for requirement, unique_thresholds in sorted(requirements.items()):
+        units, = set(u for (_, u) in unique_thresholds)  # enforces same units
+        thresh_str = ', '.join(str(v) for (v, _) in unique_thresholds)
```
Comment (TomekTrzeciak) [inline improver/wxcode/utilities.py]: Should `unique_thresholds` be sorted in the displayed extended help?

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           

Complexity decreasing per file
==============================
+ improver/wxcode/utilities.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025&bid=16191239)

File: improver_tests/cli/test_init.py [hunk 3]
```diff
@@ -200,6 +200,27 @@ def test_basic(self, m):
         self.assertEqual(m.call_count, 2)
 
 
+class Test_clizefy(unittest.TestCase):
+    """Test the clizefy decorator function"""
+
+    @patch('improver.cli.docutilize', return_value=None)
+    def test_called_once(self, m):
```
Comment (benfitzpatrick) [inline improver_tests/cli/test_init.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

File: improver_tests/cli/test_init.py [hunk 4]
```diff
@@ -200,6 +200,27 @@ def test_basic(self, m):
...
+        """Tests docutilize is called only once"""
+
+        def func():
+            """Dummy"""
+
+        clizefy(clizefy(func)).cli('argv[0]', '--help')
+        m.assert_called_once_with(func.__doc__)
+
+    @patch('improver.cli.partial')
+    def test_no_argument(self, m):
```
Comment (benfitzpatrick) [inline improver_tests/cli/test_init.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

File: improver/utilities/compare.py (hunk 1 above)
Comment (benfitzpatrick) [inline improver/utilities/compare.py]: Codacy Issue found: [Unused LOOSE_TOLERANCE imported from improver.constants](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           

Complexity decreasing per file
==============================
+ improver/wxcode/utilities.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025&bid=16191239)

File: improver/utilities/compare.py (hunk 1 above)
Comment (benfitzpatrick) [inline improver/utilities/compare.py]: Codacy Issue found: [Unused LOOSE_TOLERANCE imported from improver.constants](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

File: improver_tests/cli/test_init.py [hunk 5]
```diff
@@ -200,6 +207,27 @@ def test_basic(self, m):
         self.assertEqual(m.call_count, 2)
 
 
+class Test_clizefy(unittest.TestCase):
+    """Test the clizefy decorator function"""
+
+    @patch('improver.cli.docutilize', return_value=None)
+    def test_called_once(self, m):
```
Comment (benfitzpatrick) [inline improver_tests/cli/test_init.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

File: improver_tests/cli/test_init.py [hunk 6]
```diff
@@ -200,6 +207,27 @@ def test_basic(self, m):
...
+        """Tests docutilize is called only once"""
+
+        def func():
+            """Dummy"""
+
+        clizefy(clizefy(func)).cli('argv[0]', '--help')
+        m.assert_called_once_with(func.__doc__)
+
+    @patch('improver.cli.partial')
+    def test_no_argument(self, m):
```
Comment (benfitzpatrick) [inline improver_tests/cli/test_init.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity decreasing per file
==============================
+ improver/wxcode/utilities.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025&bid=16191239)

File: improver/utilities/compare.py (hunk 1 above)
Comment (benfitzpatrick) [inline improver/utilities/compare.py]: Codacy Issue found: [Unused LOOSE_TOLERANCE imported from improver.constants](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

File: improver/wxcode/utilities.py (hunk 2 above)
Comment (bayliffe) [reply improver/wxcode/utilities.py]: That would be good. It previously created them in the defined order, which happens to be sorted. This change means that is no longer the case, so sorting seems sensible.

File: improver/wxcode/wxcode_decision_tree.py [hunk 7]
```diff
@@ -247,8 +245,8 @@ def wxcode_decision_tree():
                 [['probability_of_lwe_snowfall_rate_above_threshold',
                   'probability_of_rainfall_rate_above_threshold']],
             'diagnostic_gamma': [1.],
-            'diagnostic_thresholds': [[AuxCoord(0.1, units='mm hr-1'),
-                                       AuxCoord(0.1, units='mm hr-1')]],
+            'diagnostic_thresholds': [[(0.1, 'mm hr-1'),
+                                       (0.1, 'mm hr-1')]],
             'diagnostic_conditions': ['above', 'above']},
```
Comment (TomekTrzeciak) [inline improver/wxcode/wxcode_decision_tree.py]: should this be `[['above', 'above']]` instead?

File: improver/wxcode/wxcode_decision_tree.py [hunk 8]
```diff
@@ -280,8 +278,8 @@ def wxcode_decision_tree():
...
```
Comment (TomekTrzeciak) [inline improver/wxcode/wxcode_decision_tree.py]: ditto

File: improver/wxcode/wxcode_decision_tree.py [hunk 9]
```diff
@@ -391,8 +389,8 @@ def wxcode_decision_tree():
                  'above_threshold',
                  'probability_of_rainfall_rate_in_vicinity_above_threshold']],
             'diagnostic_gamma': [1.],
-            'diagnostic_thresholds': [[AuxCoord(0.1, units='mm hr-1'),
-                                       AuxCoord(0.1, units='mm hr-1')]],
+            'diagnostic_thresholds': [[(0.1, 'mm hr-1'),
+                                       (0.1, 'mm hr-1')]],
             'diagnostic_conditions': ['above', 'above']},
```
Comment (TomekTrzeciak) [inline improver/wxcode/wxcode_decision_tree.py]: ditto

File: improver/wxcode/wxcode_decision_tree_global.py [hunk 10]
```diff
@@ -219,8 +217,8 @@ def wxcode_decision_tree_global():
                 [['probability_of_lwe_snowfall_rate_above_threshold',
                   'probability_of_rainfall_rate_above_threshold']],
             'diagnostic_gamma': [1.],
-            'diagnostic_thresholds': [[AuxCoord(0.1, units='mm hr-1'),
-                                       AuxCoord(0.1, units='mm hr-1')]],
+            'diagnostic_thresholds': [[(0.1, 'mm hr-1'),
+                                       (0.1, 'mm hr-1')]],
             'diagnostic_conditions': ['above', 'above']},
```
Comment (TomekTrzeciak) [inline improver/wxcode/wxcode_decision_tree_global.py]: and some more

File: improver/wxcode/wxcode_decision_tree_global.py [hunk 11]
```diff
@@ -252,8 +250,8 @@ def wxcode_decision_tree_global():
...
```
Comment (TomekTrzeciak) [inline improver/wxcode/wxcode_decision_tree_global.py]: ditto

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity decreasing per file
==============================
+ improver/wxcode/utilities.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025&bid=16191239)

File: improver/utilities/compare.py (hunk 1 above)
Comment (benfitzpatrick) [inline improver/utilities/compare.py]: Codacy Issue found: [Unused LOOSE_TOLERANCE imported from improver.constants](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4888025)

File: improver/wxcode/wxcode_decision_tree.py (hunk 7 above)
Comment (MoseleyS) [reply improver/wxcode/wxcode_decision_tree.py]: Yes, I think so (for all of these). I think the behaviour of `expand_nested_lists` is permitting these inconsistencies.

Comment (bayliffe) [reply improver/wxcode/wxcode_decision_tree.py]: There appear to be quite a few inconsistencies in how we have bracketed things. We can either fix them here, or review the lot all at once. Happy for this PR to go in unchanged and I'll open a little fix ticket to tidy up the definitions at a later date.

Comment (MoseleyS) [reply improver/wxcode/wxcode_decision_tree.py]: I agree.

**Comment by TomekTrzeciak** [comment]
I was going to fix https://github.com/metoppv/improver/pull/1132#discussion_r371739764 while I'm at it, but unfortunately it's not clear to me what's the right definition in some cases. Therefore I'm going to leave it for someone with the right expertise to fix it.

To help you out, here's the list of broken queries:
```
bad query for high_resolution tree: heavy_sleet_continuous
bad query for high_resolution tree: heavy_sleet_shower
bad query for high_resolution tree: heavy_rain_or_snow_continuous
bad query for high_resolution tree: heavy_rain_or_snow_shower
bad query for high_resolution tree: light_sleet_continuous
bad query for high_resolution tree: light_rain_or_snow_continuous
bad query for high_resolution tree: light_sleet_shower
bad query for high_resolution tree: light_rain_or_snow_shower
bad query for high_resolution tree: sleet_in_vicinity
bad query for high_resolution tree: rain_or_snow_in_vicinity
bad query for global tree: heavy_sleet_continuous
bad query for global tree: heavy_sleet_shower
bad query for global tree: heavy_rain_or_snow_continuous
bad query for global tree: heavy_rain_or_snow_shower
bad query for global tree: light_sleet_continuous
bad query for global tree: light_rain_or_snow_continuous
bad query for global tree: light_sleet_shower
bad query for global tree: light_rain_or_snow_shower
```

The above list was generated by filtering the queries against the following check:
```python
def check_query(query):
    def check_value_consistency(values):
        type_set = set(map(type, values))
        if list in type_set:
            return (
                len(type_set) == 1 and
                len(set(map(len, values))) == 1 and
                all(map(check_value_consistency, zip(*values)))
            )
        return True

    mandatory_keys = [
        'diagnostic_fields',
        'diagnostic_conditions',
        'diagnostic_thresholds',
        'probability_thresholds',
        'diagnostic_gamma',
    ]
    optional_keys = ['diagnostic_gamma']
    values_to_match = [query[key]
                       for key in mandatory_keys + optional_keys
                       if key not in optional_keys or key in query]

    return check_value_consistency(values_to_match)
```

You could stick it somewhere in wxcode plugin (or tests) to ensure consistency in the future, but instead I would encourage you to redesign the data structures used for queries. Happy to discuss further if there's interest.

**Pull request merged**